functional-tests.sh

```diff
@@ -0,0 +1,478 @@
+#   as minio server.
+#
+# * For other, call this script with environment variables MINT_MODE,
+#   MINT_DATA_DIR, SERVER_ENDPOINT, ACCESS_KEY, SECRET_KEY and ENABLE_HTTPS. It
+#   uses mc executable binary in current working directory and uses given minio
+#   server to run tests.
+#
+################################################################################
+
+if [ -n "$MINT_MODE" ]; then
+    if [ -z "${MINT_DATA_DIR+x}" ]; then
+        echo "MINT_DATA_DIR not defined"
+        exit 1
+    fi
+    if [ -z "${SERVER_ENDPOINT+x}" ]; then
+        echo "SERVER_ENDPOINT not defined"
+        exit 1
+    fi
+    if [ -z "${ACCESS_KEY+x}" ]; then
+        echo "ACCESS_KEY not defined"
+        exit 1
+    fi
+    if [ -z "${SECRET_KEY+x}" ]; then
+        echo "SECRET_KEY not defined"
+        exit 1
+    fi
+fi
+
+if [ -z "${SERVER_ENDPOINT+x}" ]; then
+    SERVER_ENDPOINT="play.minio.io:9000"
+    ACCESS_KEY="Q3AM3UQ867SPQQA43P2F"
+    SECRET_KEY="zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG"
+    ENABLE_HTTPS=1
+    SERVER_REGION="us-east-1"
+fi
+
+WORK_DIR="$PWD"
+DATA_DIR="$MINT_DATA_DIR"
+if [ -z "$MINT_MODE" ]; then
+    WORK_DIR="$PWD/.run-$RANDOM"
+    DATA_DIR="$WORK_DIR/data"
+fi
+
+FILE_1_MB="$DATA_DIR/datafile-1-MB"
+FILE_65_MB="$DATA_DIR/datafile-65-MB"
+declare FILE_1_MB_MD5SUM
+declare FILE_65_MB_MD5SUM
+
+ENDPOINT="https://$SERVER_ENDPOINT"
+if [ "$ENABLE_HTTPS" != "1" ]; then
+    ENDPOINT="http://$SERVER_ENDPOINT"
+fi
+
+SERVER_ALIAS="myminio"
+BUCKET_NAME="mc-test-bucket-$RANDOM"
+WATCH_OUT_FILE="$WORK_DIR/watch.out-$RANDOM"
+
+MC_CONFIG_DIR="/tmp/.mc-$RANDOM"
+MC="$PWD/mc"
+declare -a MC_CMD
+
+function get_md5sum()
+{
+    filename="$FILE_1_MB"
+    out=$(md5sum "$filename" 2>/dev/null)
+    rv=$?
+    if [ "$rv" -eq 0 ]; then
+        echo $(awk '{ print $1 }' <<< "$out")
+    fi
+
+    return "$rv"
+}
+
+function get_time()
+{
+    date +%s%N
+}
+
+function get_duration()
+{
+    start_time=$1
+    end_time=$(get_time)
+
+    echo $(( (end_time - start_time) / 1000000 ))
+}
+
+function log_success()
+{
+    if [ -n "$MINT_MODE" ]; then
+        printf '{"name": "mc", "duration": "%d", "function": "%s", "status": "PASS"}\n' "$(get_duration "$1")" "$2"
+    fi
+}
+
+function show()
+{
+    if [ -z "$MINT_MODE" ]; then
+        func_name="$1"
+        echo "Running $func_name()"
+    fi
+}
+
+function fail()
+{
+    rv="$1"
+    shift
+
+    if [ "$rv" -ne 0 ]; then
+        echo "$@"
+    fi
+
+    return "$rv"
+}
+
+function assert_log_failure()
+{
+    start_time="$1"
+    shift
+    func_name="$1"
+    shift
+
+    err=$("$@")
+    rv=$?
+    if [ "$rv" -ne 0 ]; then
+        if [ -n "$MINT_MODE" ]; then
+            err=$(python -c 'import sys,json; print(json.dumps(sys.stdin.read()))' <<<"$err")
+            printf '{"name": "mc", "duration": "%d", "function": "%s", "status": "FAIL", "error": "%s"}\n' "$(get_duration "$start_time")" "$func_name" "$err"
+        else
+            echo "mc: $func_name: $err"
```

Comment:
Why not print same `err` in both cases

Reply:
Because one is json friendly string, not required in another case